design-docs/configuration-dsl.md

```diff
@@ -0,0 +1,376 @@
+    myCustomGcc(Gcc) {
+        path '...'
+    }
+    myCustomVisualCpp(VisualCpp) {
+        installDir '...'
+    }
+    myCustomToolChain(ToolChain) {
+        tools.add(GccCppCompiler) {
+            it.dimensions architectures.x64, operatingSystems.windows
+            it.executable = file('/some/other/path')
+            it.withArguments { args ->
+                args << 'some-arg'
+            }
+        }
+
+        tools.add(GccLinker) {
+            it.dimensions architecture.x64, operatingSystem.windows
+            // ...
+        }
+
+        installedBy tasks.myCustomToolChainInstallTask
+
+        // TODO - Define how to add libraries
+        libraries.add(...) {
+            // added libraries can have a...
+            //   ...group, e.g. 'com.some.toolchain'
+            //   ...name, e.g. 'libc'
+            //   ...version (optional), e.g. '2.1'
+            // Adding a dependency without a version would depend on the latest or only version declared
+        }
+    }
+}
+
+repositories {
+    toolChains.myCustomToolChain
+}
+
+dependencies {
+    implementation toolChains.myCustomToolChain.libraries.libc
+    implementation visualStudio('15.0').libc
+}
+```
+
+# Build Types
+
+Build types are a conventional dimension for native binaries used to differentiate debug and release version. Additional conventional names
+are debug optimized and final. The software model, as well as Android, have this dimension as a first class citizen.
+Android typed the items inside their buildTypes container instead of relying on `ext` properties. With the new plugins, the build types will
+be typed as well.
+
+## Open Issues
+
+## DSL
+```
+apply plugin: 'build-types'
+
+buildTypes {
+    debug(MyDebugType)
+    release(BuildType)
+}
+```
+
+## Implementations
+```
+// Common build type
+public interface BuildType extends Named {}
+
+// Container type to hold build types
+public interface BuildTypeContainer extends ExtensiblePolymorphicDomainObjectContainer<BuildType> {}
+```
+
+# Linkage
+
+The linkage container specifies how you want to link your application/library/module together. The static and shared library are simple variants of
+the binary.
+
+## Open Issues
+
+* Do we want to specialize the linkage type? For example, `static(MyCustomLinkage)`?
+
+## DSL
+```
+apply plugin: 'linkage'
+
+linkages {
+    // Static library
+    static
+    // Shared library binary
+    shared
+    // Compile sources together
+    source
+    // Executable binary
+    executable
+}
+```
+
+## Implementations
+```
+public interface Linkage extends Named {}
+public interface LinkageContainer extends NamedDomainObjectContainer<Linkage> {}
+```
+
+# Prebuilt Libraries
+All native dependencies will be using Gradle's dependency management. Due to the complexity of native prebuilt library support, Gradle
+will provide a public API for declaring them.
+
+## Open Issues
+* Should Gradle allow prebuilt libraries as custom DSL or virtual repository (see example)?
```

Comment:
I guess we'll need a combination of both. For eg. if I use boost 1.62 and want to use HANA alone from 1.63, I should be able to do 

````
repositories {
    // optional block
    boost {
           hana {
               something {
                // config
            }
        }
    }
 
}

dependencies {
    implementation boost.hana
    implementation boost('1.62').asio
    implementation poco.something
}

````
This will allow configuration to those who want it and for most common users, the repository block can be completely skipped and just the dependency block should suffice. In the above case, it should auto-detect boost::asio 1.62 and [poco](https://github.com/pocoproject/poco), but use the boost.hana for HANA. 

Does that sound ok ?